File: client/lib/api/v2/schemas/agents.ts
/**
 * Agents V2 API Schemas
 * Schema definitions for agents v2 endpoints
 */

import { z } from "zod";
import {
  DepartmentMappingSchema,
  ModelMappingSchema,
  ReasoningMappingSchema,
} from "./base";

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const AgentsListRequestSchema = z.object({
  profileId: z.string(),
});

export type AgentsListRequest = z.infer<typeof AgentsListRequestSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Agent item for list view
export const AgentItemSchema = z.object({
  agent_id: z.string(),
  name: z.string(),
  description: z.string(),
  reasoning: z.string().nullable(),
  temperature: z.number(),
  model_id: z.string(),
  role: z.string(),
  department_ids: z.array(z.string()).nullable(),
  updated_at: z.string(),
  can_edit: z.boolean(),
  can_duplicate: z.boolean(),
  can_delete: z.boolean(),
});

export const AgentsListResponseSchema = z.object({
  agents: z.array(AgentItemSchema),
  model_mapping: ModelMappingSchema,
  department_mapping: DepartmentMappingSchema,
});

export type AgentsListResponse = z.infer<typeof AgentsListResponseSchema>;
export type AgentItem = z.infer<typeof AgentItemSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

export const AgentDetailRequestSchema = z.object({
  agentId: z.string(),
  profileId: z.string(),
});

export type AgentDetailRequest = z.infer<typeof AgentDetailRequestSchema>;

export const AgentDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type AgentDetailDefaultRequest = z.infer<
  typeof AgentDetailDefaultRequestSchema
>;

// Debug info item
export const DebugInfoItemSchema = z.object({
  created_at: z.string(),
  model_id: z.string(),
  content: z.string(),
});

export type DebugInfoItem = z.infer<typeof DebugInfoItemSchema>;

export const AgentDetailResponseSchema = z.object({
  // Basic fields
  name: z.string(),
  description: z.string(),
  system_prompt: z.string(),
  prompt_id: z.string().nullable(),
  temperature: z.number(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  active: z.boolean(),
  role: z.string(), // agent_role enum value

  // Prompt version history
  prompt_mapping: z.record(
    z.string(),
    z.object({
      system_prompt: z.string(),
      created_at: z.string(),
      updated_at: z.string(),
      department_ids: z.array(z.string()).nullable(),
    })
  ),

  // Department associations
  department_ids: z.array(z.string()),
  valid_department_ids: z.array(z.string()),
  department_mapping: DepartmentMappingSchema,

  // Metadata
  valid_model_ids: z.array(z.string()),
  reasoning_options: z.array(z.string()),
  temperature_lower: z.number(),
  temperature_upper: z.number(),

  // Debug info
  debug_info: z.array(DebugInfoItemSchema),

  // Mappings
  model_mapping: ModelMappingSchema,
  reasoning_mapping: ReasoningMappingSchema,
});

export type AgentDetailResponse = z.infer<typeof AgentDetailResponseSchema>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Create request
export const CreateAgentRequestSchema = z.object({
  name: z.string(),
  description: z.string(),
  prompt_id: z.string().nullable().optional(), // If provided, use existing prompt
  system_prompt: z.string(), // If prompt_id is None, create new prompt with this
  temperature: z.number(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  active: z.boolean(),
  role: z.string(), // agent_role enum value
  department_ids: z.array(z.string()).nullable(), // None = cross-department (superadmin only)
});

export type CreateAgentRequest = z.infer<typeof CreateAgentRequestSchema>;

export const CreateAgentResponseSchema = z.object({
  success: z.boolean(),
  agentId: z.string(),
  message: z.string(),
});

export type CreateAgentResponse = z.infer<typeof CreateAgentResponseSchema>;

// Update request
export const UpdateAgentRequestSchema = z.object({
  agentId: z.string(),
  name: z.string(),
  description: z.string(),
  prompt_id: z.string().nullable().optional(), // If provided, use existing prompt
  system_prompt: z.string(), // If prompt_id is None, create new prompt with this
  temperature: z.number(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  active: z.boolean(),
  role: z.string(), // agent_role enum value
  department_ids: z.array(z.string()).nullable(), // None = cross-department (superadmin only)
});

export type UpdateAgentRequest = z.infer<typeof UpdateAgentRequestSchema>;

export const UpdateAgentResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type UpdateAgentResponse = z.infer<typeof UpdateAgentResponseSchema>;

// Duplicate request
export const DuplicateAgentRequestSchema = z.object({
  agentId: z.string(),
});

export type DuplicateAgentRequest = z.infer<typeof DuplicateAgentRequestSchema>;

export const DuplicateAgentResponseSchema = z.object({
  success: z.boolean(),
  agentId: z.string(),
  message: z.string(),
});

export type DuplicateAgentResponse = z.infer<
  typeof DuplicateAgentResponseSchema
>;

// Delete request
export const DeleteAgentRequestSchema = z.object({
  agentId: z.string(),
});

export type DeleteAgentRequest = z.infer<typeof DeleteAgentRequestSchema>;

export const DeleteAgentResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeleteAgentResponse = z.infer<typeof DeleteAgentResponseSchema>;

File: server/app/schemas/agents.py
"""Agents V2 API schemas."""

from pydantic import BaseModel

from .base import DepartmentMapping, ModelMapping, ReasoningMapping

# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class AgentsListRequest(BaseModel):
    """Request for agents list."""

    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class AgentItem(BaseModel):
    """Agent item for list view."""

    agent_id: str
    name: str
    description: str
    reasoning: str | None
    temperature: float
    model_id: str
    role: str
    department_ids: list[str] | None
    updated_at: str
    can_edit: bool
    can_duplicate: bool
    can_delete: bool


class AgentsListResponse(BaseModel):
    """Response for agents list."""

    agents: list[AgentItem]
    model_mapping: ModelMapping
    department_mapping: DepartmentMapping


# ============================================================================
# DETAIL SCHEMAS
# ============================================================================


class AgentDetailRequest(BaseModel):
    """Request for agent detail."""

    agentId: str
    profileId: str


class AgentDetailDefaultRequest(BaseModel):
    """Request for default agent detail."""

    profileId: str


class DebugInfoItem(BaseModel):
    """Debug information item."""

    created_at: str
    model_id: str
    content: str


class PromptInfo(BaseModel):
    """Prompt information for version history."""

    system_prompt: str
    created_at: str
    updated_at: str
    department_ids: list[str] | None


class AgentDetailResponse(BaseModel):
    """Response for agent detail."""

    # Basic fields
    name: str
    description: str
    system_prompt: str
    prompt_id: str | None
    temperature: float
    model_id: str
    reasoning: str | None
    active: bool
    role: str  # agent_role enum value

    # Metadata
    valid_model_ids: list[str]
    reasoning_options: list[str]
    temperature_lower: float
    temperature_upper: float

    # Department associations
    department_ids: list[str]
    valid_department_ids: list[str]
    department_mapping: DepartmentMapping

    # Prompt version history
    prompt_mapping: dict[str, PromptInfo]

    # Debug info
    debug_info: list[DebugInfoItem]

    # Mappings
    model_mapping: ModelMapping
    reasoning_mapping: ReasoningMapping


# ============================================================================
# MUTATION SCHEMAS
# ============================================================================


class CreateAgentRequest(BaseModel):
    """Request for creating an agent."""

    name: str
    description: str
    prompt_id: str | None  # If provided, use existing prompt
    system_prompt: str  # If prompt_id is None, create new prompt with this
    temperature: float
    model_id: str
    reasoning: str | None
    active: bool
    role: str  # agent_role enum value
    department_ids: list[str] | None  # None = cross-department (superadmin only)


class CreateAgentResponse(BaseModel):
    """Response for creating an agent."""

    success: bool
    agentId: str
    message: str


class UpdateAgentRequest(BaseModel):
    """Request for updating an agent."""

    agentId: str
    name: str
    description: str
    prompt_id: str | None  # If provided, use existing prompt
    system_prompt: str  # If prompt_id is None, create new prompt with this
    temperature: float
    model_id: str
    reasoning: str | None
    active: bool
    role: str  # agent_role enum value
    department_ids: list[str] | None  # None = cross-department (superadmin only)


class UpdateAgentResponse(BaseModel):
    """Response for updating an agent."""

    success: bool
    message: str


class DuplicateAgentRequest(BaseModel):
    """Request for duplicating an agent."""

    agentId: str


class DuplicateAgentResponse(BaseModel):
    """Response for duplicating an agent."""

    success: bool
    agentId: str
    message: str


class DeleteAgentRequest(BaseModel):
    """Request for deleting an agent."""

    agentId: str


class DeleteAgentResponse(BaseModel):
    """Response for deleting an agent."""

    success: bool
    message: str
